jupyterlab_gcedetails/src/service/price_service.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ServerConnection } from '@jupyterlab/services';
+import {
+  HardwareConfiguration,
+  REGIONS,
+  HOURS_PER_MONTH,
+  NO_SUSTAINED_DISCOUNT_PREFIXES,
+} from '../data/data';
+
+/** The monthly price catalog for region and sku */
+type Catalog = Map<string, Map<string, number>>;
+
+export class PriceService {
+  private catalog: Catalog;
+  private _receivedError: boolean;
+
+  private async buildCatalog(): Promise<void> {
+    // Only attempt to build the catalog once
+    if (!this.catalog) {
```

Comment:
Just a nit, I'd suggest inverting this and short-circuit right away if this.catalog is populated rather than having to have the entire method body inside of an if block.